core/src/avm2/globals/flash/net/socket.rs

```diff
@@ -0,0 +1,365 @@
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        return Ok(socket
+            .read_float()
+            .map_err(|e| e.to_avm(activation))?
+            .into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn read_int<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        return Ok(socket.read_int().map_err(|e| e.to_avm(activation))?.into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn read_short<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        return Ok(socket
+            .read_short()
+            .map_err(|e| e.to_avm(activation))?
+            .into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn read_unsigned_byte<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        return Ok(socket
+            .read_unsigned_byte()
+            .map_err(|e| e.to_avm(activation))?
+            .into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn read_unsigned_int<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        return Ok(socket
+            .read_unsigned_int()
+            .map_err(|e| e.to_avm(activation))?
+            .into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn read_unsigned_short<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        return Ok(socket
+            .read_unsigned_short()
+            .map_err(|e| e.to_avm(activation))?
+            .into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn write_byte<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        let byte = args.get_i32(activation, 0)?;
+        socket.write_bytes(&[byte as u8]);
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn write_bytes<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(socket) = this.as_socket() {
+        let bytearray = args.get_object(activation, 0, "bytes")?;
+        let offset = args.get_u32(activation, 1)? as usize;
+        let length = args.get_u32(activation, 2)? as usize;
+
+        let ba_read = bytearray
+            .as_bytearray()
+            .ok_or("ArgumentError: Parameter must be a bytearray")?;
```

Comment:
This is guaranteed to be a ByteArray by the AS-side bindings and the `get_object` filtering out any `nill`s, so you can just do `.expect("Parameter must be a bytearray!")`.